rfc-149-switch-to-per-page-asset-loading.md

```diff
@@ -0,0 +1,208 @@
+
+Supplying the assets that a component needs as part of the partial would mean that assets would only be included when the component is used - this would simplify the developer experience when adding or removing components from applications, and would also mean a better user experience since unused assets wouldn't be loaded.
+
+## Proposal
+
+Assets should be included individually on a per-component basis and included in the component partial.
+
+This would mean that a page loads `component.css` and `component.js` only when a component is present on the page - rather than the current behaviour, which sees component assets loaded if used anywhere within an application. If multiple uses of the same component occur on the page, there should only be one instance of the assets included on the page.
+
+### Keep the stylesheets in the head
+
+Placing stylesheets in the `head` will ensure that all of the required CSS is downloaded and parsed before rendering begins. This is the current behaviour, so shouldn't be changed.
+
+Whilst putting a `link rel="stylesheet"` into the body is [okay][link_stylesheet_is_body_okay], this would lead to too many changes at once, too many unknowns, and isn't considered best practice. If this proposal is implemented, further investigation could be carried to find out which is the better way of loading assets to improve painting performance.
+
+### Keep JavaScript files just before the closing body tag
+
+[Slimmer][slimmer] currently takes all `script` elements that don't have `defer` or `async` attributes and [moves them to the end of the markup][slimmer_moves_scripts]. This ensures that they are parsed and executed in a known order when the DOM is ready to be read. This behaviour shouldn't be changed.
+
+### Shared assets
+
+The assets should be placed in a folder that allows each application to reference them. The current behaviour has each application place their assets in a different folder - `/assets/collections`, `/assets/finder-frontend` etc. This prevents the browser cache from being used effectively as `/assets/collection/accordion.css` will be downloaded even if `/assets/finder-frontend/accordion.css` has already been downloaded.
+
+A version identifier should be added to the component asset filename to help identify which version of the components gem is being used - for example `gem-v29-2-1-accordion.css`. Similarly, the bespoke app assets should be renamed to their application - so collections would have `collections.css` and `collections.js` - to make it easier to see which stylesheet comes from which application.
+
+The assets would ideally be placed in a shared folder as per [RFC #91][sharing_assets], so would be able to be used across GOV.UK irrespective of the rendering application. Since this hasn't been implemented yet, there would need to be some work for this to happen - and likely this would be delayed to coincide with the replatforming work.
```

Comment:
I'll rewrite this to be clearer - but the tl;dr is that post RFC-91 and RFC-149 the asset URLs should look like:

* `www.gov.uk/assets/gem-v29-2-1-accordion-e03sv...98hg5.css`
* `www.gov.uk/assets/government-frontend-00d69...a20c7.css`

(The assets are currently served off of the `www.gov.uk` domain since [HTTP/2 was enabled](https://github.com/alphagov/govuk-rfcs/blob/main/rfc-115-enabling-http2-on-govuk.md#proposal).)

The renaming of the `application-*.css` is because I think (though am happy to be wrong) that Rails can only use a single asset folder, which is set using `config.assets.prefix`[^1]:

```ruby
    # Path within public/ where assets are compiled to
    config.assets.prefix = "/assets/frontend"
``` 

If this is correct, then a Rails app wouldn't be able to use assets from `/shared` and `/government-frontend` at the same time.

Since all the applications use `application-*.css` and `application-*.js` for their bespoke assets, there'd be a lot of similarly named files where the only difference is the fingerprint - which isn't that useful when trying to tell which file is used by which application. Setting the file name to include the application name seems a good solution to that.

I'd prefer to keep the fingerprint as part of the individual assets to make doubly sure that each file with the same name has the same contents - for example, the same Sass file compiled with Dart-Sass would give a different fingerprint to one compiled with Node-Sass because they have different default number of decimal places[^2].



[^1]: Example taken from [alphagov/frontend/config/application.rb](https://github.com/alphagov/frontend/blob/3ccc1bf52c980852eca61250727f81034930519d/config/application.rb#L58-L59).
[^2]: There's an [interesting issue on how this affects GOV.UK Frontend](https://github.com/alphagov/govuk-frontend/issues/2199).

Reply:
Thanks - I think I mostly follow. Can we look at an example just to iron out the last of the confusion?

If we take Frontend as an example and look at its [use of the tabs component](https://github.com/alphagov/frontend/blob/09b27f918b987afb6b1f6dddfdbfb8961dbadcd0/app/views/calendar/bank_holidays.html.erb#L69), it currently [imports the SCSS](https://github.com/alphagov/frontend/blob/6ec5131c8edeceff4e60b118576482be805a9e40/app/assets/stylesheets/application.scss#L35) and [JS](https://github.com/alphagov/frontend/blob/8dbe2e4de266be77da511156f476fb52b2b09166/app/assets/javascripts/application.js#L7) manually, combining it with all of the application assets and outputting as single files `/assets/frontend/application-FINGERPRINT.css` and `/assets/frontend/application-FINGERPRINT.js` respectively.

For stage 1, we want to implement the individual serving of assets. This would involve:

1. Removing the `import`/`require` calls to the component
2. Pulling in the component CSS/JS in some other way - **this is where I'm stuck**. It may be an implementation detail but it could end up changing parts of the proposal, so worth digging a little deeper I think. I'm not clear from the RFC how this would be done.

I think what you're proposing is roughly this:

> Write a compile-time 'hook' into the `render` method, which copies the component CSS/JS files into the Rails project, as separate files, alongside the application asset source files.
> This will then output as `/assets/frontend/gem-v29-2-1-tabs-FINGERPRINT.css` etc. We rely on Slimmer to yank the `<link rel` element and move it into the head.
> The component template itself makes no reference to assets.
>
> Then for stage 2, it's just a case of changing the `config.assets.prefix` value to something like `/assets/shared` and renaming `application.scss` and `application.js` to `frontend.scss|js`. _Now_ the component would be uploaded to `/assets/shared/gem-v29-2-1-tabs-FINGERPRINT.css` instead.
> And if another app (e.g. Whitehall) uses the same version of the same component, it will also output to the same place, essentially overwriting the contents, but we don't care since the contents (and therefore the fingerprint) would be the same. So at that point, Frontend is effectively using the tabs assets compiled by Whitehall, but there is no difference in the bytes being sent to the user.

Does that make sense? Have I understood that right?

The other alternative I can think of is to have the component templates refer to their own assets, like in the pseudocode below, but I can't think of a way of accomplishing that whilst retaining the fingerprint, and it would also need a separate build pipeline (likely a GitHub Action).

```erb
<% if use_individual_assets %>
  <link rel="stylesheet" href="<%=`/assets/shared/gem-v${VERSION}-tabs.css` %>" />
<% end %>
``` 

(Might make the 'optional' logic easier though - just a case of passing `use_individual_assets: false` at the application level).

Reply:
Yes, that's what I think should happen. There's some discussion that should happen during the implementation about exactly how and where things should happen - but that's getting a bit too much into the weeds. 

From digging into how the component gem interacts with the rendering applications I can't find anything that screams that this isn't doable. GitHub have started to break their CSS bundle up into more managable chunks - [there was a presentation at this years RailsConf on it](https://youtu.be/21QG19Zy_g0?t=989) ([slides](https://hawksley.org/2022/03/01/conference-proposal-breaking-up-with-the-bundle.html)) - which also makes me think that this is a achievable.

Reply:
I've rewritten the section in https://github.com/alphagov/govuk-rfcs/pull/152/commits/adbc60a9ab8861811d3098f5a4c7dea78b552681 - does it make more sense?